src/app/ui/renderer.js

```diff
@@ -76,48 +76,72 @@ function Renderer (appAPI) {
   }
 }
 
-Renderer.prototype.error = function (message, container, options) {
+/**
+ * format msg like error or warning,
+ *
+ * @param {String or DOMElement} message
+ * @param {DOMElement} container
+ * @param {Object} options {useSpan, noAnnotations, click:(Function), type:(warning, error), errFile, errLine, errCol}
+ */
+Renderer.prototype.error = function (message, container, opt) {
   if (container === undefined) return
-  var self = this
-  var opt = options || {}
+  opt = opt || {}
+
+  var text
+  if (typeof message === 'string') {
+    text = message
+    message = yo`<span>${message}</span>`
+  } else if (message.innerText) {
+    text = message.innerText
+  }
+
   if (!opt.type) {
-    opt.type = utils.errortype(message)
+    opt.type = utils.errortype(text)
   }
-  var $pre
-  if (opt.isHTML) {
-    $pre = $(opt.useSpan ? '<span />' : '<pre />').html(message)
-  } else {
-    $pre = $(opt.useSpan ? '<span />' : '<pre />').text(message)
+
+  var errLocation = text.match(/^([^:]*):([0-9]*):(([0-9]*):)? /)
+  if (errLocation) {
+    errLocation = parseRegExError(errLocation)
+    opt.errFile = errLocation.errFile
+    opt.errLine = errLocation.errLine
+    opt.errCol = errLocation.errCol
   }
-  var $error = $('<div class="sol ' + opt.type + '"><div class="close"><i class="fa fa-close"></i></div></div>').prepend($pre)
-  container.append($error)
-  var err = message.match(/^([^:]*):([0-9]*):(([0-9]*):)? /)
-  if (err) {
-    var errFile = err[1]
-    var errLine = parseInt(err[2], 10) - 1
-    var errCol = err[4] ? parseInt(err[4], 10) : 0
-    if (!opt.noAnnotations) {
-      self.appAPI.error(errFile, {
-        row: errLine,
-        column: errCol,
-        text: message,
-        type: opt.type
-      })
-    }
-    $error.click(function (ev) {
-      options && options.click ? options.click(errFile, errLine, errCol) : self.appAPI.errorClick(errFile, errLine, errCol)
-    })
-  } else if (options && options.click) {
-    $error.click(function (ev) {
-      options.click(message)
+
+  if (!opt.noAnnotations && errLocation) {
+    this.appAPI.error(errLocation.errFile, {
+      row: errLocation.errLine,
+      column: errLocation.errCol,
+      text: text,
+      type: opt.type
     })
   }
 
+  var $pre = $(opt.useSpan ? '<span />' : '<pre />').html(message)
```

Comment:
Is there no way to do that with yoyo?

Reply:
yes but didn't want to change everything, and this part is not going to be harmful
I can use yoyo here though